Add FNV hash implementation

codope opened the pull request:
Implements [Issue 114](https://github.com/airlift/slice/issues/114).
Essentially, there are two flavors of the hash function; `hash32` and `hash64`

martint commented on src/main/java/io/airlift/slice/FNVHash.java:
```diff
@@ -0,0 +1,56 @@
+    public static int hash32(Slice data)
+    {
+        int hash = FNV_32_OFFSET_BASIS;
+        final byte[] dataBytes = data.getBytes();
```
getBytes() makes a copy of the underlying bytes. Consider using `getByte(int index)` instead. See how we did it for other hashes (e.g., https://github.com/airlift/slice/blob/master/src/main/java/io/airlift/slice/SpookyHashV2.java#L159).

martint commented on src/test/java/io/airlift/slice/TestFNVHash.java:
```diff
@@ -0,0 +1,80 @@
+    @BeforeTest
+    public void setup()
+    {
+        testData = new byte[][] {
```
Splitting the data from the expected values makes the code harder to follow and maintain. I would suggest either using TestNG data providers or inlining this into each of the test methods that use it.

martint commented on src/test/java/io/airlift/slice/TestFNVHash.java:
```diff
@@ -0,0 +1,80 @@
+                "foobar".getBytes(),
+                "foobar!\n".getBytes(),
+                "line 1\nline 2\nline 3".getBytes(),
+                "http://en.wikipedia.org/wiki/Fowler_Noll_Vo_hash".getBytes(),
```
`String.getBytes()` uses the default platform encoding, so it might produce different results in different environments. Use something like `getBytes(StandardCharsets.UTF_8)`

martint commented on src/main/java/io/airlift/slice/FNVHash.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Reference implementation: https://tools.ietf.org/html/draft-eastlake-fnv-17#section-6
+ */
+public class FNVHash
```
Capitalize as `FnvHash`. We do that to improve readability when many acronyms are used in a row (e.g., `HTTPURL` vs `HttpUrl`, `HTTPXMLRPC` vs `HttpXmlRpc`).

martint commented on src/main/java/io/airlift/slice/FNVHash.java:
```diff
@@ -0,0 +1,56 @@
+        int hash = FNV_32_OFFSET_BASIS;
+        final byte[] dataBytes = data.getBytes();
+
+        for (final byte dataByte : dataBytes) {
```
No need for `final` in local variables.

codope commented:
@martint Hey Martin, thanks for review this. I have addressed all the comments. Could you please review again?

codope commented:
@martint , I have addressed the comments. Could you please review and merge the diff? A Presto issue is waiting on this merge; https://github.com/prestodb/presto/issues/12500#issuecomment-507653908

martint commented:
Sorry for the delay. Thanks!

Are you also planning to contribute the implementation of the functions to the community version of Presto? (https://github.com/prestosql/presto)

codope commented:
Yes, I'll try to add it to community version by this week.

codope commented:
@martint I want to add the function to Presto and use the latest version of slice. However, [maven central](https://mvnrepository.com/artifact/io.airlift/slice) is still showing v0.36. Can you please release the latest version? I'll update the pom in Presto accordingly.

widagdos commented on src/main/java/io/airlift/slice/FnvHash.java:
```diff
@@ -0,0 +1,56 @@
+        int hash = FNV_32_OFFSET_BASIS;
+
+        for (int i = 0; i < data.length(); ++i) {
+            byte dataByte = data.getByte(i);
```
getByte should be replaced with getUnsignedByte. Otherwise when the byte value msb is set (the value is negative), there will be sign extension on the next line.
